### PR TITLE
[ re #258 ] Utility functions on N-ary products

### PR DESCRIPTION
**Pull request by gallais**

### CODE REVIEW

**Comment by MatthewDaggitt** [comment]
Great! Though I think the proofs below `unzip` should be moved to `Data.Product.N-ary.Properties`...

**Comment by MatthewDaggitt** [comment]
> So I'll move ↔Vec : ∀ {a} {A : Set a} n → A ^ n I.↔ Vec A n there too?
That's not backwards compatible (but probably OK: given how empty the
module was, I'd be surprised if anyone actually used it).

Yes I think so. The alternative is that the main file becomes bigger and bigger and we'll end up needing to create a `.Properties` at a later date which will cause many more backwards compatibility issues. Better to nip it in the bud now.

We can view it as part of the rite-of-passage of actually becoming a useful module!

File: src/Data/Product/N-ary.agda [hunk 1]
```diff
@@ -12,55 +12,136 @@
...
+
+ zipWith : (A → B → C) → ∀ n → A ^ n → B ^ n → C ^ n
+ zipWith f 0      as       bs       = []
+ zipWith f 1      a        b        = f a b
+ zipWith f (2+ n) (a , as) (b , bs) = f a b , zipWith f (suc n) as bs
+
+ unzipWith : (A → B × C) → ∀ n → A ^ n → B ^ n × C ^ n
+ unzipWith f 0      as       = [] , []
+ unzipWith f 1      a        = f a
+ unzipWith f (2+ n) (a , as) =
```
Comment (MatthewDaggitt) [inline src/Data/Product/N-ary.agda]: This is much neater written as `P.zip _,_ _,_ (f a) (unzipWith f (suc n) as)`

**Comment by MatthewDaggitt** [comment]
Great, thanks. The only other issue I've noticed is that you've only indented the code inside the anonymous modules by a single space instead of the standard two spaces we use for indentation everywhere else in the library. You've done the same in #270. Was this a deliberate choice or just an oversight?

**Comment by gallais** [comment]
This is something I have gotten used to in my own code for the content of
anonymous modules. Didn't even notice I was doing it here!

**Comment by MatthewDaggitt** [comment]
> This is something I have gotten used to in my own code for the content of
anonymous modules. Didn't even notice I was doing it here!

Thanks, felt slightly petty even bringing it up!

**Pull request merged**